<div style="
    width:100%;
    box-sizing:border-box;
    background:
        radial-gradient(circle at 12% 12%,rgba(194,139,88,0.10),transparent 24%),
        radial-gradient(circle at 88% 82%,rgba(111,125,91,0.09),transparent 25%),
        linear-gradient(135deg,#FAF7F1 0%,#FFFDF9 52%,#F3F0E8 100%);
    padding:36px 40px 32px 40px;
    border-radius:30px;
    font-family:'Segoe UI',Arial,sans-serif;
    border:1px solid #DDD4C4;
    box-shadow:0 14px 34px rgba(91,76,53,0.10);
    overflow:hidden;
">

<div style="
    display:flex;
    align-items:center;
    justify-content:space-between;
    margin-bottom:24px;
">

<div style="
    color:#8A6245;
    font-size:13px;
    font-weight:900;
    letter-spacing:3px;
">
WEEK 6 • DAY 4
</div>

<div style="
    color:#5E684F;
    font-size:11px;
    font-weight:900;
    letter-spacing:1.5px;
    padding:8px 14px;
    border:1px solid #C9CBB9;
    border-radius:18px;
    background:#F1F2E8;
">
DEEP LEARNING • KERAS
</div>

</div>


<div style="
    text-align:center;
">

<div style="
    color:#4A4338;
    font-size:14px;
    font-weight:900;
    letter-spacing:2.5px;
    margin-bottom:12px;
">
FROM ARCHITECTURE TO TRAINING
</div>

<div style="
    color:#2F2A24;
    font-size:46px;
    font-weight:900;
    line-height:1.05;
    letter-spacing:-1.4px;
">
Building & Training
</div>

<div style="
    color:#9A6947;
    font-size:30px;
    font-weight:900;
    line-height:1.15;
    margin-top:6px;
">
a Neural Network in Keras
</div>

<div style="
    width:100px;
    height:5px;
    margin:20px auto 17px auto;
    background:#B97852;
    border-radius:20px;
">
</div>

<div style="
    max-width:780px;
    margin:0 auto;
    color:#6B665D;
    font-size:15px;
    line-height:1.8;
    font-weight:600;
">
Build the network layer by layer, compile it with the right
optimizer and loss, train it through epochs, and use the training
history to understand how the model learns.
</div>

</div>


<div style="
    width:100%;
    max-width:880px;
    height:310px;
    margin:28px auto 0 auto;
    box-sizing:border-box;
    position:relative;
">

<div style="
    position:absolute;
    inset:0;
    background:#E9E3D8;
    border-radius:25px;
    transform:rotate(-1deg);
">
</div>

<div style="
    position:relative;
    width:100%;
    height:100%;
    box-sizing:border-box;
    background:#FFFDFC;
    border:1px solid #D8CEBD;
    border-radius:25px;
    box-shadow:0 12px 26px rgba(91,76,53,0.10);
    overflow:hidden;
    display:flex;
    align-items:center;
    justify-content:center;
">

<div style="
    position:absolute;
    top:15px;
    left:17px;
    z-index:2;
    color:#7B563F;
    background:#F6EEE4;
    border:1px solid #E2CDBA;
    padding:7px 12px;
    border-radius:15px;
    font-size:10px;
    font-weight:900;
    letter-spacing:1.3px;
">
MODEL ARCHITECTURE
</div>

<img src="../media/keras_neural_network.jpg"
     alt="Keras neural network architecture"
     style="
        display:block;
        max-width:94%;
        max-height:250px;
        width:auto;
        height:auto;
        object-fit:contain;
        margin-top:10px;
     ">

</div>

</div>


<div style="
    margin-top:25px;
    display:grid;
    grid-template-columns:1fr 1fr 1fr;
    gap:14px;
    width:100%;
    box-sizing:border-box;
">


<div style="
    width:100%;
    box-sizing:border-box;
    background:#F5EFE6;
    border:1px solid #E0D0BE;
    border-radius:16px;
    padding:18px 15px;
    text-align:center;
">

<div style="
    color:#9A6947;
    font-size:11px;
    font-weight:900;
    letter-spacing:1.3px;
">
01 • ARCHITECTURE
</div>

<div style="
    color:#4A4338;
    font-size:15px;
    font-weight:900;
    margin-top:7px;
">
Sequential + Dense
</div>

</div>


<div style="
    width:100%;
    box-sizing:border-box;
    background:#EFF1E8;
    border:1px solid #D0D4C2;
    border-radius:16px;
    padding:18px 15px;
    text-align:center;
">

<div style="
    color:#687455;
    font-size:11px;
    font-weight:900;
    letter-spacing:1.3px;
">
02 • TRAINING
</div>

<div style="
    color:#4A4F3E;
    font-size:15px;
    font-weight:900;
    margin-top:7px;
">
Compile + Fit
</div>

</div>


<div style="
    width:100%;
    box-sizing:border-box;
    background:#F9F0E9;
    border:1px solid #E4C6B1;
    border-radius:16px;
    padding:18px 15px;
    text-align:center;
">

<div style="
    color:#A35F3D;
    font-size:11px;
    font-weight:900;
    letter-spacing:1.3px;
">
03 • GENERALIZATION
</div>

<div style="
    color:#684535;
    font-size:15px;
    font-weight:900;
    margin-top:7px;
">
BatchNorm + Dropout
</div>

</div>

</div>


<div style="
    margin-top:20px;
    padding:16px 20px;
    background:#F7F4EE;
    border:1px dashed #CFC5B4;
    border-radius:15px;
    text-align:center;
">

<span style="
    color:#8A6245;
    font-size:11px;
    font-weight:900;
    letter-spacing:1.4px;
">
TODAY'S IDEA
</span>

<span style="
    color:#6B665D;
    font-size:13px;
    font-weight:600;
    margin-left:9px;
">
Keras handles the mathematics so we can focus on
</span>

<span style="
    color:#5E684F;
    font-size:13px;
    font-weight:900;
">
designing, training, and improving the model.
</span>

</div>


<div style="
    margin-top:18px;
    text-align:center;
">

<span style="
    color:#9A9186;
    font-size:10px;
    font-weight:900;
    letter-spacing:1.6px;
">
TENSORFLOW • KERAS • TRAINING HISTORY • REGULARIZATION
</span>

</div>

</div>

<div style="
    width:100%;
    box-sizing:border-box;
    background:
        radial-gradient(circle at 92% 10%,rgba(111,125,91,0.08),transparent 24%),
        linear-gradient(135deg,#FAF7F1 0%,#FFFDF9 55%,#F3F0E8 100%);
    padding:26px 30px;
    border-radius:24px;
    font-family:'Segoe UI',Arial,sans-serif;
    border:1px solid #DDD4C4;
    box-shadow:0 9px 24px rgba(91,76,53,0.07);
    overflow:hidden;
">

<div style="
    display:flex;
    align-items:center;
    gap:18px;
">

<div style="
    flex-shrink:0;
    color:#FFFDF9;
    background:linear-gradient(145deg,#9A6947,#81543A);
    width:92px;
    height:92px;
    border-radius:22px;
    display:flex;
    align-items:center;
    justify-content:center;
    font-size:38px;
    font-weight:900;
    letter-spacing:-1.5px;
    box-shadow:0 8px 18px rgba(154,105,71,0.18);
">
4.1
</div>

<div style="
    flex:1;
    min-width:0;
">

<div style="
    color:#687455;
    font-size:10px;
    font-weight:900;
    letter-spacing:2.2px;
    margin-bottom:6px;
">
THE FRAMEWORK
</div>

<div style="
    color:#2F2A24;
    font-size:30px;
    font-weight:900;
    line-height:1.15;
">
TensorFlow / Keras
</div>

<div style="
    color:#6B665D;
    font-size:14px;
    font-weight:600;
    line-height:1.8;
    margin-top:7px;
    max-width:900px;
">
The high-level deep learning framework that lets us define
neural networks layer by layer while handling
<span style="
    color:#5E684F;
    font-weight:900;
">
forward propagation, backpropagation, and optimization
</span>
automatically.
</div>

</div>

</div>


<div style="
    margin-top:18px;
    padding-top:14px;
    border-top:1px solid #DDD4C4;
    display:flex;
    align-items:center;
    justify-content:space-between;
    gap:18px;
">

<div style="
    color:#9A6947;
    font-size:10px;
    font-weight:900;
    letter-spacing:1.4px;
">
HIGH-LEVEL API
</div>

<div style="
    color:#687455;
    font-size:11px;
    font-weight:800;
    text-align:right;
">
Build the model → Keras handles the math
</div>

</div>

</div>

<div style="
    width:100%;
    box-sizing:border-box;
    background:
        radial-gradient(circle at 92% 8%,rgba(111,125,91,0.07),transparent 24%),
        linear-gradient(135deg,#FAF7F1 0%,#FFFDF9 55%,#F3F0E8 100%);
    padding:30px 34px;
    border-radius:24px;
    font-family:'Segoe UI',Arial,sans-serif;
    border:1px solid #DDD4C4;
    box-shadow:0 9px 24px rgba(91,76,53,0.07);
    overflow:hidden;
">


<div style="
    color:#2F2A24;
    font-size:30px;
    font-weight:900;
    line-height:1.2;
">
Keras as the High-Level Framework
</div>

<div style="
    width:100px;
    height:5px;
    background:#B97852;
    border-radius:20px;
    margin:16px 0 17px 0;
">
</div>

<div style="
    color:#6B665D;
    font-size:16px;
    font-weight:600;
    line-height:1.9;
    max-width:1000px;
">
Keras provides a high-level interface for building neural networks.
Instead of implementing the mathematical operations manually, we describe
the architecture layer by layer and the framework handles the training process
automatically.
</div>


<div style="
    margin-top:24px;
    display:grid;
    grid-template-columns:repeat(3,minmax(0,1fr));
    gap:15px;
    width:100%;
    box-sizing:border-box;
">


<div style="
    width:100%;
    min-width:0;
    box-sizing:border-box;
    background:#F5EFE6;
    border:1px solid #E0D0BE;
    border-radius:16px;
    padding:19px 17px;
">

<div style="
    color:#9A6947;
    font-size:12px;
    font-weight:900;
    letter-spacing:1.2px;
">
01 • FORWARD PROPAGATION
</div>

<div style="
    color:#4A4338;
    font-size:15px;
    font-weight:800;
    line-height:1.6;
    margin-top:8px;
">
Moves the input through the network to produce a prediction.
</div>

</div>


<div style="
    width:100%;
    min-width:0;
    box-sizing:border-box;
    background:#EFF1E8;
    border:1px solid #D0D4C2;
    border-radius:16px;
    padding:19px 17px;
">

<div style="
    color:#687455;
    font-size:12px;
    font-weight:900;
    letter-spacing:1.2px;
">
02 • BACKPROPAGATION
</div>

<div style="
    color:#4A4F3E;
    font-size:15px;
    font-weight:800;
    line-height:1.6;
    margin-top:8px;
">
Calculates gradients to determine how the weights should change.
</div>

</div>


<div style="
    width:100%;
    min-width:0;
    box-sizing:border-box;
    background:#F9F0E9;
    border:1px solid #E4C6B1;
    border-radius:16px;
    padding:19px 17px;
">

<div style="
    color:#A35F3D;
    font-size:12px;
    font-weight:900;
    letter-spacing:1.2px;
">
03 • OPTIMIZATION
</div>

<div style="
    color:#684535;
    font-size:15px;
    font-weight:800;
    line-height:1.6;
    margin-top:8px;
">
Updates the weights to reduce the loss and improve predictions.
</div>

</div>

</div>


<div style="
    margin-top:21px;
    padding:17px 20px;
    background:#F7F4EE;
    border:1px dashed #CFC5B4;
    border-radius:15px;
    text-align:center;
">

<span style="
    color:#687455;
    font-size:12px;
    font-weight:900;
    letter-spacing:1.3px;
">
KEY IDEA
</span>

<span style="
    color:#6B665D;
    font-size:15px;
    font-weight:600;
    margin-left:10px;
">
You define
</span>

<span style="
    color:#9A6947;
    font-size:15px;
    font-weight:900;
">
what the network looks like
</span>

<span style="
    color:#6B665D;
    font-size:15px;
    font-weight:600;
">
and Keras handles
</span>

<span style="
    color:#5E684F;
    font-size:15px;
    font-weight:900;
">
how it learns.
</span>

</div>

</div>

<div style="
    width:100%;
    box-sizing:border-box;
    background:
        linear-gradient(135deg,#FFFDF9 0%,#F4F0E8 100%);
    padding:26px 32px;
    border-radius:22px;
    font-family:'Segoe UI',Arial,sans-serif;
    border:1px solid #DDD4C4;
    box-shadow:0 8px 22px rgba(91,76,53,0.06);
">



<div style="
    color:#2F2A24;
    font-size:30px;
    font-weight:900;
    line-height:1.2;
">
PyTorch
</div>

<div style="
    width:90px;
    height:5px;
    background:#687455;
    border-radius:20px;
    margin:15px 0 16px 0;
">
</div>

<div style="
    color:#6B665D;
    font-size:16px;
    line-height:1.9;
    font-weight:600;
    max-width:1000px;
">
PyTorch is another major deep learning framework and the main alternative to
TensorFlow/Keras. It provides tools for building, training, and evaluating
neural networks, but
<span style="
    color:#9A6947;
    font-weight:900;
">
it is not used in today's implementation.
</span>
</div>

<div style="
    margin-top:18px;
    padding:15px 18px;
    background:#F0F1E8;
    border:1px solid #D0D4C2;
    border-radius:14px;
">

<span style="
    color:#687455;
    font-size:12px;
    font-weight:900;
    letter-spacing:1.2px;
">
TODAY
</span>

<span style="
    color:#6B665D;
    font-size:15px;
    font-weight:600;
    margin-left:9px;
">
We will use
</span>

<span style="
    color:#9A6947;
    font-size:15px;
    font-weight:900;
">
TensorFlow / Keras
</span>

<span style="
    color:#6B665D;
    font-size:15px;
    font-weight:600;
">
for the models and experiments in this notebook.
</span>

</div>

</div>

<div style="
    width:100%;
    box-sizing:border-box;
    background:
        radial-gradient(circle at 92% 10%,rgba(111,125,91,0.08),transparent 24%),
        linear-gradient(135deg,#FAF7F1 0%,#FFFDF9 55%,#F3F0E8 100%);
    padding:26px 30px;
    border-radius:24px;
    font-family:'Segoe UI',Arial,sans-serif;
    border:1px solid #DDD4C4;
    box-shadow:0 9px 24px rgba(91,76,53,0.07);
    overflow:hidden;
">

<div style="
    display:flex;
    align-items:center;
    gap:18px;
">

<div style="
    flex-shrink:0;
    color:#FFFDF9;
    background:linear-gradient(145deg,#9A6947,#81543A);
    width:92px;
    height:92px;
    border-radius:22px;
    display:flex;
    align-items:center;
    justify-content:center;
    font-size:38px;
    font-weight:900;
    letter-spacing:-1.5px;
    box-shadow:0 8px 18px rgba(154,105,71,0.18);
">
4.2
</div>

<div style="
    flex:1;
    min-width:0;
">

<div style="
    color:#687455;
    font-size:10px;
    font-weight:900;
    letter-spacing:2.2px;
    margin-bottom:6px;
">
MODEL ARCHITECTURE
</div>

<div style="
    color:#2F2A24;
    font-size:30px;
    font-weight:900;
    line-height:1.15;
">
Building a Model with the Sequential API
</div>

<div style="
    color:#6B665D;
    font-size:14px;
    font-weight:600;
    line-height:1.8;
    margin-top:7px;
    max-width:900px;
">
The Sequential API builds a neural network by stacking layers in order,
while Dense layers create fully connected connections between neurons.
</div>

</div>

</div>


<div style="
    margin-top:18px;
    padding-top:14px;
    border-top:1px solid #DDD4C4;
    display:flex;
    align-items:center;
    justify-content:space-between;
    gap:18px;
">

<div style="
    color:#9A6947;
    font-size:10px;
    font-weight:900;
    letter-spacing:1.4px;
">
SEQUENTIAL API
</div>

<div style="
    color:#687455;
    font-size:11px;
    font-weight:800;
    text-align:right;
">
Stack layers → Connect neurons → Learn patterns
</div>

</div>

</div>

<div style="
    width:100%;
    box-sizing:border-box;
    background:#FFFDF9;
    padding:30px 34px;
    border-radius:22px;
    font-family:'Segoe UI',Arial,sans-serif;
    border:1px solid #DDD4C4;
    box-shadow:0 8px 22px rgba(91,76,53,0.06);
">

<div style="
    color:#687455;
    font-size:14px;
    font-weight:900;
    letter-spacing:1.7px;
    margin-bottom:8px;
">
HOW THE MODEL IS BUILT
</div>

<div style="
    color:#2F2A24;
    font-size:28px;
    font-weight:900;
    line-height:1.2;
">
Sequential API & Dense Layers
</div>

<div style="
    color:#6B665D;
    font-size:16px;
    line-height:1.9;
    font-weight:600;
    margin-top:12px;
">
The
<span style="color:#9A6947;font-weight:900;">Sequential API</span>
builds a neural network by stacking layers in a specific order.
Each layer receives the output of the previous layer and passes its own
output to the next one.
</div>

<div style="
    margin-top:20px;
    padding:18px 20px;
    background:#F5EFE6;
    border:1px solid #E0D0BE;
    border-radius:16px;
">

<div style="
    color:#9A6947;
    font-size:13px;
    font-weight:900;
    letter-spacing:1.2px;
    margin-bottom:8px;
">
DENSE = FULLY CONNECTED
</div>

<div style="
    color:#4A4338;
    font-size:15px;
    line-height:1.8;
    font-weight:700;
">
A
<span style="color:#9A6947;font-weight:900;">Dense</span>
layer connects every neuron in one layer to every neuron in the next layer.
It learns weights and a bias for these connections and applies the selected
activation function to produce its output.
</div>

</div>

<div style="
    margin-top:18px;
    display:grid;
    grid-template-columns:repeat(3,minmax(0,1fr));
    gap:14px;
    width:100%;
    box-sizing:border-box;
">

<div style="
    background:#F5EFE6;
    border:1px solid #E0D0BE;
    border-radius:15px;
    padding:16px;
    box-sizing:border-box;
">

<div style="
    color:#9A6947;
    font-size:12px;
    font-weight:900;
">
LAYER 1
</div>

<div style="
    color:#4A4338;
    font-size:18px;
    font-weight:900;
    margin-top:6px;
">
64 neurons
</div>

<div style="
    color:#6B665D;
    font-size:13px;
    font-weight:600;
    margin-top:4px;
">
ReLU activation
</div>

</div>

<div style="
    background:#EFF1E8;
    border:1px solid #D0D4C2;
    border-radius:15px;
    padding:16px;
    box-sizing:border-box;
">

<div style="
    color:#687455;
    font-size:12px;
    font-weight:900;
">
LAYER 2
</div>

<div style="
    color:#4A4F3E;
    font-size:18px;
    font-weight:900;
    margin-top:6px;
">
32 neurons
</div>

<div style="
    color:#6B665D;
    font-size:13px;
    font-weight:600;
    margin-top:4px;
">
ReLU activation
</div>

</div>

<div style="
    background:#F9F0E9;
    border:1px solid #E4C6B1;
    border-radius:15px;
    padding:16px;
    box-sizing:border-box;
">

<div style="
    color:#A35F3D;
    font-size:12px;
    font-weight:900;
">
OUTPUT
</div>

<div style="
    color:#684535;
    font-size:18px;
    font-weight:900;
    margin-top:6px;
">
1 neuron
</div>

<div style="
    color:#6B665D;
    font-size:13px;
    font-weight:600;
    margin-top:4px;
">
Sigmoid activation
</div>

</div>

</div>

</div>

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

df = pd.read_csv("heart_disease_uci.csv", sep=";")

X = df.drop(columns=["cardio"])
y = df["cardio"]

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

print("Dataset shape:", df.shape)
print("Training shape:", X_train.shape)
print("Test shape:", X_test.shape)
print("Target classes:", sorted(y.unique()))

Dataset shape: (70000, 13)
Training shape: (56000, 12)
Test shape: (14000, 12)
Target classes: [np.int64(0), np.int64(1)]


In [3]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

input_dim = X_train.shape[1]

model = keras.Sequential([
    layers.Input(shape=(input_dim,)),
    layers.Dense(64, activation="relu"),
    layers.Dense(32, activation="relu"),
    layers.Dense(1, activation="sigmoid")
])

model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 64)             │           832 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 1)              │            33 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 2,945 (11.50 KB)

 Trainable params: 2,945 (11.50 KB)

 Non-trainable params: 0 (0.00 B)

<div style="
    width:100%;
    box-sizing:border-box;
    background:#FFFDF9;
    padding:22px 26px;
    border-radius:19px;
    font-family:'Segoe UI',Arial,sans-serif;
    border:1px solid #DDD4C4;
">

<div style="
    color:#9A6947;
    font-size:12px;
    font-weight:900;
    letter-spacing:1.6px;
">
MODEL SUMMARY
</div>

<div style="
    color:#2F2A24;
    font-size:22px;
    font-weight:900;
    margin-top:6px;
">
A Compact Binary Classification Network
</div>

<div style="
    color:#6B665D;
    font-size:14px;
    line-height:1.8;
    font-weight:600;
    margin-top:8px;
">
The network contains three Dense layers with ReLU activations in the hidden layers
and a Sigmoid output for binary classification. The model has
<span style="color:#9A6947;font-weight:900;">
2,945 trainable parameters
</span>
that will be optimized during training.
</div>

</div>

<div style="
    width:100%;
    box-sizing:border-box;
    background:
        radial-gradient(circle at 92% 10%,rgba(111,125,91,0.08),transparent 24%),
        linear-gradient(135deg,#FAF7F1 0%,#FFFDF9 55%,#F3F0E8 100%);
    padding:26px 30px;
    border-radius:24px;
    font-family:'Segoe UI',Arial,sans-serif;
    border:1px solid #DDD4C4;
    box-shadow:0 9px 24px rgba(91,76,53,0.07);
    overflow:hidden;
">

<div style="
    display:flex;
    align-items:center;
    gap:18px;
">

<div style="
    flex-shrink:0;
    color:#FFFDF9;
    background:linear-gradient(145deg,#9A6947,#81543A);
    width:92px;
    height:92px;
    border-radius:22px;
    display:flex;
    align-items:center;
    justify-content:center;
    font-size:38px;
    font-weight:900;
    letter-spacing:-1.5px;
    box-shadow:0 8px 18px rgba(154,105,71,0.18);
">
4.3
</div>

<div style="
    flex:1;
    min-width:0;
">

<div style="
    color:#687455;
    font-size:10px;
    font-weight:900;
    letter-spacing:2.2px;
    margin-bottom:6px;
">
TRAINING WORKFLOW
</div>

<div style="
    color:#2F2A24;
    font-size:30px;
    font-weight:900;
    line-height:1.15;
">
Compile, Train, Evaluate
</div>

<div style="
    color:#6B665D;
    font-size:14px;
    font-weight:600;
    line-height:1.8;
    margin-top:7px;
    max-width:900px;
">
The standard Keras workflow moves from configuring the model,
to training it on data, and finally evaluating its performance
on unseen test samples.
</div>

</div>

</div>


<div style="
    margin-top:18px;
    padding-top:14px;
    border-top:1px solid #DDD4C4;
    display:grid;
    grid-template-columns:1fr 25px 1fr 25px 1fr;
    align-items:center;
    gap:8px;
">

<div style="
    background:#F5EFE6;
    border:1px solid #E0D0BE;
    border-radius:13px;
    padding:11px;
    text-align:center;
">

<div style="color:#9A6947;font-size:10px;font-weight:900;">
01
</div>

<div style="color:#4A4338;font-size:12px;font-weight:900;margin-top:4px;">
COMPILE
</div>

</div>

<div style="
    color:#9A6947;
    font-size:18px;
    font-weight:900;
    text-align:center;
">
→
</div>

<div style="
    background:#EFF1E8;
    border:1px solid #D0D4C2;
    border-radius:13px;
    padding:11px;
    text-align:center;
">

<div style="color:#687455;font-size:10px;font-weight:900;">
02
</div>

<div style="color:#4A4F3E;font-size:12px;font-weight:900;margin-top:4px;">
FIT
</div>

</div>

<div style="
    color:#687455;
    font-size:18px;
    font-weight:900;
    text-align:center;
">
→
</div>

<div style="
    background:#F9F0E9;
    border:1px solid #E4C6B1;
    border-radius:13px;
    padding:11px;
    text-align:center;
">

<div style="color:#A35F3D;font-size:10px;font-weight:900;">
03
</div>

<div style="color:#684535;font-size:12px;font-weight:900;margin-top:4px;">
EVALUATE
</div>

</div>

</div>

</div>

<div style="
    width:100%;
    box-sizing:border-box;
    background:#FFFDF9;
    padding:28px 32px;
    border-radius:22px;
    font-family:'Segoe UI',Arial,sans-serif;
    border:1px solid #DDD4C4;
    box-shadow:0 8px 22px rgba(91,76,53,0.06);
    overflow:hidden;
">

<div style="
    color:#9A6947;
    font-size:13px;
    font-weight:900;
    letter-spacing:1.8px;
    margin-bottom:8px;
">
FROM WEEK 3 TO KERAS
</div>

<div style="
    color:#2F2A24;
    font-size:27px;
    font-weight:900;
    line-height:1.2;
">
The Same ML Logic — A Different Framework
</div>

<div style="
    color:#6B665D;
    font-size:16px;
    line-height:1.9;
    font-weight:600;
    margin-top:11px;
">
The Keras workflow follows the same basic idea we used with
Scikit-learn in Week 3. We first configure the model, then train it
on the training data, and finally measure its performance on unseen data.
</div>


<!-- COMPILE -->

<div style="
    width:100%;
    box-sizing:border-box;
    margin-top:22px;
    padding:17px 19px;
    background:#F5EFE6;
    border:1px solid #E0D0BE;
    border-left:5px solid #9A6947;
    border-radius:15px;
">

<div style="
    color:#9A6947;
    font-size:13px;
    font-weight:900;
    letter-spacing:1.2px;
">
01 • COMPILE
</div>

<div style="
    color:#4A4338;
    font-size:15px;
    font-weight:800;
    line-height:1.7;
    margin-top:6px;
">
Set the optimizer, loss function, and evaluation metrics before training begins.
</div>

</div>


<!-- FIT -->

<div style="
    width:100%;
    box-sizing:border-box;
    margin-top:12px;
    padding:17px 19px;
    background:#EFF1E8;
    border:1px solid #D0D4C2;
    border-left:5px solid #687455;
    border-radius:15px;
">

<div style="
    color:#687455;
    font-size:13px;
    font-weight:900;
    letter-spacing:1.2px;
">
02 • FIT
</div>

<div style="
    color:#4A4F3E;
    font-size:15px;
    font-weight:800;
    line-height:1.7;
    margin-top:6px;
">
Run the training loop across batches and epochs using the training data.
A validation split can be used to monitor generalization during training.
</div>

</div>


<!-- EVALUATE -->

<div style="
    width:100%;
    box-sizing:border-box;
    margin-top:12px;
    padding:17px 19px;
    background:#F9F0E9;
    border:1px solid #E4C6B1;
    border-left:5px solid #A35F3D;
    border-radius:15px;
">

<div style="
    color:#A35F3D;
    font-size:13px;
    font-weight:900;
    letter-spacing:1.2px;
">
03 • EVALUATE
</div>

<div style="
    color:#684535;
    font-size:15px;
    font-weight:800;
    line-height:1.7;
    margin-top:6px;
">
Measure the final performance on data the model has never seen during training.
</div>

</div>


<!-- COMPARISON -->

<div style="
    width:100%;
    box-sizing:border-box;
    margin-top:19px;
    padding:16px 19px;
    background:#F7F4EE;
    border:1px dashed #CFC5B4;
    border-radius:14px;
    text-align:center;
">

<div style="
    color:#9A6947;
    font-size:12px;
    font-weight:900;
    letter-spacing:1.2px;
    margin-bottom:7px;
">
SCIKIT-LEARN → KERAS
</div>

<div style="
    color:#6B665D;
    font-size:15px;
    font-weight:700;
    line-height:1.7;
">

<span style="color:#4A4338;font-weight:900;">
Week 3:
</span>

fit / predict

<span style="color:#B97852;font-size:20px;font-weight:900;margin:0 9px;">
→
</span>

<span style="color:#5E684F;font-weight:900;">
Day 4:
</span>

compile / fit / evaluate

</div>

</div>

</div>

In [4]:
model.compile(
    optimizer="adam",
    loss="binary_crossentropy",
    metrics=["accuracy"]
)

print("Model compiled successfully.")

Model compiled successfully.


In [5]:
history = model.fit(
    X_train,
    y_train,
    validation_split=0.20,
    epochs=50,
    batch_size=32,
    verbose=1
)

Epoch 1/50
1400/1400 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.6830 - loss: 0.5995 - val_accuracy: 0.7271 - val_loss: 0.5730
Epoch 2/50
1400/1400 ━━━━━━━━━━━━━━━━━━━━ 7s 3ms/step - accuracy: 0.7227 - loss: 0.5665 - val_accuracy: 0.7352 - val_loss: 0.5592
Epoch 3/50
1400/1400 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7266 - loss: 0.5566 - val_accuracy: 0.7362 - val_loss: 0.5482
Epoch 4/50
1400/1400 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7314 - loss: 0.5495 - val_accuracy: 0.7396 - val_loss: 0.5450
Epoch 5/50
1400/1400 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7314 - loss: 0.5464 - val_accuracy: 0.7363 - val_loss: 0.5475
Epoch 6/50
1400/1400 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7303 - loss: 0.5445 - val_accuracy: 0.7371 - val_loss: 0.5417
Epoch 7/50
1400/1400 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7347 - loss: 0.5433 - val_accuracy: 0.7370 - val_loss: 0.5409
Epoch 8/50
1400/1400 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.7334 - loss: 0.5426 -

In [6]:
test_loss, test_accuracy = model.evaluate(
    X_test,
    y_test,
    verbose=0
)

print(f"Test Loss: {test_loss:.4f}")
print(f"Test Accuracy: {test_accuracy:.2%}")

Test Loss: 0.5563
Test Accuracy: 72.71%


<div style="
    width:100%;
    box-sizing:border-box;
    background:
        linear-gradient(135deg,#FFFDF9 0%,#F3F0E8 100%);
    padding:28px 32px;
    border-radius:22px;
    font-family:'Segoe UI',Arial,sans-serif;
    border:1px solid #DDD4C4;
    box-shadow:0 8px 22px rgba(91,76,53,0.06);
    overflow:hidden;
">

<div style="
    color:#687455;
    font-size:13px;
    font-weight:900;
    letter-spacing:1.7px;
    margin-bottom:7px;
">
READING THE TRAINING HISTORY
</div>

<div style="
    color:#2F2A24;
    font-size:28px;
    font-weight:900;
    line-height:1.2;
">
What Does the `history` Object Store?
</div>

<div style="
    color:#6B665D;
    font-size:16px;
    line-height:1.9;
    font-weight:600;
    margin-top:12px;
">
The object returned by
<span style="color:#9A6947;font-weight:900;">
model.fit()
</span>
stores the loss and metric values recorded after every epoch.
Because we used a validation split, it keeps both training and validation results.
</div>


<!-- TRAINING -->

<div style="
    width:100%;
    box-sizing:border-box;
    margin-top:21px;
    padding:18px 20px;
    background:#F5EFE6;
    border:1px solid #E0D0BE;
    border-left:5px solid #9A6947;
    border-radius:15px;
">

<div style="
    color:#9A6947;
    font-size:13px;
    font-weight:900;
    letter-spacing:1.2px;
">
TRAINING VALUES
</div>

<div style="
    color:#4A4338;
    font-size:15px;
    line-height:1.8;
    font-weight:700;
    margin-top:7px;
">
`loss` and `accuracy` are recorded for every epoch using the training data.
These values show how the model performs while it is learning from the training set.
</div>

</div>


<!-- VALIDATION -->

<div style="
    width:100%;
    box-sizing:border-box;
    margin-top:12px;
    padding:18px 20px;
    background:#EFF1E8;
    border:1px solid #D0D4C2;
    border-left:5px solid #687455;
    border-radius:15px;
">

<div style="
    color:#687455;
    font-size:13px;
    font-weight:900;
    letter-spacing:1.2px;
">
VALIDATION VALUES
</div>

<div style="
    color:#4A4F3E;
    font-size:15px;
    line-height:1.8;
    font-weight:700;
    margin-top:7px;
">
`val_loss` and `val_accuracy` are recorded for the validation portion of the data.
They help us understand how well the model generalizes to data it is not directly learning from.
</div>

</div>


<!-- WHY IT MATTERS -->

<div style="
    width:100%;
    box-sizing:border-box;
    margin-top:18px;
    padding:16px 19px;
    background:#F9F0E9;
    border:1px solid #E4C6B1;
    border-radius:14px;
    text-align:center;
">

<div style="
    color:#A35F3D;
    font-size:12px;
    font-weight:900;
    letter-spacing:1.2px;
    margin-bottom:7px;
">
WHY IT MATTERS
</div>

<div style="
    color:#6B665D;
    font-size:14px;
    font-weight:700;
    line-height:1.7;
">
Plotting these values helps us diagnose
<span style="color:#684535;font-weight:900;">
overfitting, underfitting, and generalization
</span>
throughout training.
</div>

</div>

</div>

<div style="
    width:100%;
    box-sizing:border-box;
    background:
        radial-gradient(circle at 92% 10%,rgba(111,125,91,0.08),transparent 24%),
        linear-gradient(135deg,#FAF7F1 0%,#FFFDF9 55%,#F3F0E8 100%);
    padding:26px 30px;
    border-radius:24px;
    font-family:'Segoe UI',Arial,sans-serif;
    border:1px solid #DDD4C4;
    box-shadow:0 9px 24px rgba(91,76,53,0.07);
    overflow:hidden;
">

<div style="
    display:flex;
    align-items:center;
    gap:18px;
">

<div style="
    flex-shrink:0;
    color:#FFFDF9;
    background:linear-gradient(145deg,#9A6947,#81543A);
    width:92px;
    height:92px;
    border-radius:22px;
    display:flex;
    align-items:center;
    justify-content:center;
    font-size:38px;
    font-weight:900;
    letter-spacing:-1.5px;
    box-shadow:0 8px 18px rgba(154,105,71,0.18);
">
4.4
</div>

<div style="
    flex:1;
    min-width:0;
">

<div style="
    color:#687455;
    font-size:10px;
    font-weight:900;
    letter-spacing:2.2px;
    margin-bottom:6px;
">
REGULARIZATION & STABILITY
</div>

<div style="
    color:#2F2A24;
    font-size:30px;
    font-weight:900;
    line-height:1.15;
">
Batch Normalization and Dropout
</div>

<div style="
    color:#6B665D;
    font-size:14px;
    font-weight:600;
    line-height:1.8;
    margin-top:7px;
    max-width:900px;
">
Two practical techniques that help neural networks train more
stably and reduce overfitting.
</div>

</div>

</div>

</div>

<div style="
    width:100%;
    box-sizing:border-box;
    background:#FFFDF9;
    padding:28px 32px;
    border-radius:22px;
    font-family:'Segoe UI',Arial,sans-serif;
    border:1px solid #DDD4C4;
    box-shadow:0 8px 22px rgba(91,76,53,0.06);
    overflow:hidden;
">

<div style="
    color:#9A6947;
    font-size:13px;
    font-weight:900;
    letter-spacing:1.8px;
    margin-bottom:8px;
">
BATCH NORMALIZATION
</div>

<div style="
    color:#2F2A24;
    font-size:28px;
    font-weight:900;
    line-height:1.2;
">
Make Training More Stable
</div>

<div style="
    color:#6B665D;
    font-size:16px;
    line-height:1.9;
    font-weight:600;
    margin-top:12px;
">
Batch Normalization normalizes the activations flowing through a layer during training.
This helps keep the values on a more consistent scale, which can make training
<span style="color:#9A6947;font-weight:900;">
faster, more stable, and easier to optimize
</span>.
</div>

<div style="
    margin-top:20px;
    width:100%;
    box-sizing:border-box;
    padding:17px 20px;
    background:#F5EFE6;
    border:1px solid #E0D0BE;
    border-left:5px solid #B97852;
    border-radius:15px;
">

<div style="
    color:#9A6947;
    font-size:12px;
    font-weight:900;
    letter-spacing:1.2px;
">
MAIN BENEFIT
</div>

<div style="
    color:#4A4338;
    font-size:15px;
    font-weight:800;
    line-height:1.8;
    margin-top:6px;
">
More stable activations → smoother optimization → more reliable training.
</div>

</div>

</div>

<div style="
    width:100%;
    box-sizing:border-box;
    background:
        linear-gradient(135deg,#FFFDF9 0%,#F9F0E9 100%);
    padding:28px 32px;
    border-radius:22px;
    font-family:'Segoe UI',Arial,sans-serif;
    border:1px solid #E4C6B1;
    box-shadow:0 8px 22px rgba(91,76,53,0.06);
    overflow:hidden;
">

<div style="
    color:#A35F3D;
    font-size:13px;
    font-weight:900;
    letter-spacing:1.8px;
    margin-bottom:8px;
">
DROPOUT
</div>

<div style="
    color:#2F2A24;
    font-size:28px;
    font-weight:900;
    line-height:1.2;
">
Reduce Overfitting
</div>

<div style="
    color:#6B665D;
    font-size:16px;
    line-height:1.9;
    font-weight:600;
    margin-top:12px;
">
Dropout randomly switches off a fraction of neurons during each training step.
This prevents the network from depending too heavily on specific neurons or paths
and encourages it to learn more robust representations.
</div>

<div style="
    margin-top:20px;
    width:100%;
    box-sizing:border-box;
    padding:17px 20px;
    background:#F5EFE6;
    border:1px solid #E0D0BE;
    border-left:5px solid #9A6947;
    border-radius:15px;
">

<div style="
    color:#9A6947;
    font-size:12px;
    font-weight:900;
    letter-spacing:1.2px;
">
CONNECTION TO WEEK 4
</div>

<div style="
    color:#4A4338;
    font-size:15px;
    font-weight:800;
    line-height:1.8;
    margin-top:6px;
">
Dropout acts as a regularization technique, similar in purpose to the
regularization methods used to control model complexity and improve generalization.
</div>

</div>

</div>

In [11]:
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense, BatchNormalization, Dropout


improved_model = Sequential([
    Dense(64, activation="relu", input_shape=(input_dim,)),
    BatchNormalization(),
    Dropout(0.3),

    Dense(32, activation="relu"),
    BatchNormalization(),
    Dropout(0.3),

    Dense(1, activation="sigmoid")
])

from IPython.display import display, HTML

rows = []

for layer in improved_model.layers:
    output_shape = tuple(layer.output.shape)
    params = layer.count_params()

    rows.append(f"""
    <tr>
        <td style="
            padding:13px 16px;
            text-align:left;
            vertical-align:middle;
            border-bottom:1px solid #E5D8CB;
            color:#4A4338;
            font-weight:800;
        ">
            {layer.name}
        </td>

        <td style="
            padding:13px 16px;
            text-align:left;
            vertical-align:middle;
            border-bottom:1px solid #E5D8CB;
            color:#684535;
            font-weight:700;
        ">
            {layer.__class__.__name__}
        </td>

        <td style="
            padding:13px 16px;
            text-align:left;
            vertical-align:middle;
            border-bottom:1px solid #E5D8CB;
            color:#5E684F;
            font-weight:700;
        ">
            {output_shape}
        </td>

        <td style="
            padding:13px 16px;
            text-align:left;
            vertical-align:middle;
            border-bottom:1px solid #E5D8CB;
            color:#9A6947;
            font-weight:900;
        ">
            {params:,}
        </td>
    </tr>
    """)

total_params = improved_model.count_params()

display(HTML(f"""
<div style="
    width:100%;
    box-sizing:border-box;
    background:#FFFDF9;
    padding:24px 26px;
    border-radius:20px;
    font-family:'Segoe UI',Arial,sans-serif;
    border:1px solid #DDD4C4;
    box-shadow:0 8px 22px rgba(91,76,53,0.07);
    overflow:hidden;
">

<div style="
    color:#9A6947;
    font-size:12px;
    font-weight:900;
    letter-spacing:1.8px;
    margin-bottom:6px;
    text-align:left;
">
MODEL SUMMARY
</div>

<div style="
    color:#2F2A24;
    font-size:25px;
    font-weight:900;
    margin-bottom:18px;
    text-align:left;
">
Improved Neural Network
</div>

<table style="
    width:100%;
    border-collapse:collapse;
    table-layout:auto;
    border:1px solid #D8C6B7;
    border-radius:14px;
    overflow:hidden;
    font-size:13px;
    text-align:left;
">

<thead>

<tr style="
    background:#8B5E3C;
    color:#FFFDF9;
">

<th style="
    padding:14px 16px;
    text-align:left;
    vertical-align:middle;
    font-weight:900;
    white-space:nowrap;
">
Layer
</th>

<th style="
    padding:14px 16px;
    text-align:left;
    vertical-align:middle;
    font-weight:900;
    white-space:nowrap;
">
Type
</th>

<th style="
    padding:14px 16px;
    text-align:left;
    vertical-align:middle;
    font-weight:900;
    white-space:nowrap;
">
Output Shape
</th>

<th style="
    padding:14px 16px;
    text-align:left;
    vertical-align:middle;
    font-weight:900;
    white-space:nowrap;
">
Parameters
</th>

</tr>

</thead>

<tbody>

{''.join(rows)}

</tbody>

</table>

<div style="
    margin-top:15px;
    padding:14px 17px;
    background:#F5EFE6;
    border:1px solid #E0D0BE;
    border-radius:13px;
    display:flex;
    align-items:center;
    justify-content:space-between;
    text-align:left;
">

<div style="
    color:#9A6947;
    font-size:11px;
    font-weight:900;
    letter-spacing:1.2px;
    text-align:left;
">
TOTAL PARAMETERS
</div>

<div style="
    color:#684535;
    font-size:19px;
    font-weight:900;
    text-align:left;
">
{total_params:,}
</div>

</div>

</div>
"""))

Layer,Type,Output Shape,Parameters
dense_15,Dense,"(None, 64)",832
batch_normalization_8,BatchNormalization,"(None, 64)",256
dropout_8,Dropout,"(None, 64)",0
dense_16,Dense,"(None, 32)","2,080"
batch_normalization_9,BatchNormalization,"(None, 32)",128
dropout_9,Dropout,"(None, 32)",0
dense_17,Dense,"(None, 1)",33
